File: 0566_reshape_the_matrix/Solution.java
// LeetCode 0566 - Reshape the Matrix
// https://leetcode.com/problems/reshape-the-matrix/

class Solution {
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        int rows = mat.length;
        int cols = mat[0].length;
        if (rows * cols != r * c) {
            return mat;
        }

        int[][] result = new int[r][c];
        int index = 0;
        for (int i = 0; i < r; ++i) {
            for (int j = 0; j < c; ++j) {
                result[i][j] = mat[index / cols][index % cols];
                ++index;
            }
        }
        return result;
    }
}

File: 0566_reshape_the_matrix/Solution.scala
// LeetCode 0566 - Reshape the Matrix
// https://leetcode.com/problems/reshape-the-matrix/

object Solution {
  def matrixReshape(mat: Array[Array[Int]], r: Int, c: Int): Array[Array[Int]] = {
    val rows = mat.length
    val cols = mat(0).length
    if (rows * cols != r * c) return mat
    val result = Array.ofDim[Int](r, c)
    var index = 0
    var i = 0
    while (i < r) {
      var j = 0
      while (j < c) {
        result(i)(j) = mat(index / cols)(index % cols)
        index += 1
        j += 1
      }
      i += 1
    }
    result
  }
}
